examples/instructor-gemini-flash/application.py

```diff
@@ -0,0 +1,226 @@
+            topic_str += f"SUBTOPIC {i + 1}: {subtopic.name}\n"
+            for j, concept in enumerate(subtopic.concepts):
+                topic_str += f"CONCEPT {j + 1}: {concept}\n"
+        return topic_str
+
+
+def topics_system_template(
+    num_required_topics: int = NUM_REQUIRED_TOPICS,
+    min_words: int = MIN_WORDS,
+    max_words: int = MAX_WORDS,
+    min_subtopics: int = MIN_SUBTOPICS,
+    max_subtopics: int = MAX_SUBTOPICS,
+    min_concepts: int = MIN_CONCEPTS,
+    max_concepts: int = MAX_CONCEPTS,
+) -> str:
+    return f"""\
+"You are a world class course instructor."
+You'll be given a course outline and you have to generate {num_required_topics} topics.
+For each topic:
+    1. Generate a {min_words}-{max_words} word topic name that encapsulates the description.
+    2. Generate {min_subtopics}-{max_subtopics} subtopics for the topic. Also {min_words}-{max_words} words each.
+    For each subtopic:
+        Generate {min_concepts}-{max_concepts} concepts. Also {min_words}-{max_words} words each. The concepts should be related to the subtopic.
+        Think of concepts as the smallest unit of knowledge that can be taught from the subtopic. And add a verb to the concept to make it actionable.
+        For example:
+            "Calculate Derivatives" instead of "Derivatives".
+            "Identify Finite Sets" instead of "Finite Sets".
+            "Find the y-intercept" instead of "y-intercept".
+    The subtopics and concepts should be in the correct order.\
+"""
+
+
+def creation_template(
+    outline: str,
+    topics_so_far: list[Topic],
+    num_required_topics: int = NUM_REQUIRED_TOPICS,
+) -> str:
+    prompt_strs = [
+        f"<outline>\n{outline}\n</outline>",
+        f"<num_required_topics>\n{num_required_topics}\n</num_required_topics>",
+    ]
+    topics_so_far_str = "<topics_so_far>\n"
+    if topics_so_far:
+        for i, topic in enumerate(topics_so_far):
+            topics_so_far_str += f"{i+1}/{num_required_topics}\n{topic}\n"
+    topics_so_far_str = topics_so_far_str.strip() + "\n</topics_so_far>"
+    prompt_strs.append(topics_so_far_str)
+    prompt_strs.append("Generate the next topic.")
+    return "\n\n".join(prompt_strs)
+
+
+@action(reads=[], writes=["outline", "num_required_topics", "chat_history"])
+def setup(state: State, outline: str, num_required_topics: int) -> State:
+    return state.update(
+        outline=outline,
+        num_required_topics=num_required_topics,
+        chat_history=[
+            {
+                "role": "system",
+                "content": topics_system_template(
+                    num_required_topics=num_required_topics
+                ),
+            }
+        ],
+    )
+
+
+@action(
+    reads=[
+        "chat_history",
+        "outline",
+        "num_required_topics",
+        "topics_so_far",
+        "topic_feedback",
+    ],
+    writes=["generated_topic", "chat_history"],
+)
+def creator(
```

Comment:
Can you add docstring comments?

Reply:
Mind adding just short docstring comments? But to each one? E.G. the first sentence of this, but for each action. Makes it easier to read/undersatnd the example :)

Reply:
Of course. My bad. Added.

Reply:
Oh, specifically I meant docstrings to the functions decorated by `@action` -- These are the important ones that tell the story. Think you have enough comments in this function!

Reply:
Updated :memo: